toolbox/HBTReader.py

```diff
@@ -1,279 +1,686 @@
+            (HostHaloId, m0, len(trackIds), isnap))
+
+        hosts = []
+        for trackId in trackIds:
+            track = self.GetTrack(trackId, MaxSnap=isnap)
+            logging.debug("Track %d of halo %d@%d across %d snapshots" %
+                          (trackId, HostHaloId, isnap, len(track)))
+            hosts_of_track = zip(track['Snapshot'], track['HostHaloId'])
+            hosts.extend(hosts_of_track)
+
+        hosts = np.unique(
+            np.array(
+                hosts, dtype=np.dtype([('Snapshot', int), ('HaloId', int)])),
+            axis=0)
+        hosts = hosts[hosts['HaloId'] != 0]
+
+        masses = [
+            self.GetHostHalo(host['HaloId'], host['Snapshot'])['M200Crit']
+            for host in hosts
+        ]
+        hosts = append_fields(hosts, 'M200Crit', masses, usemask=False)
+        snaps = np.unique(hosts['Snapshot'])
+
+        logging.debug(
+            "Queried halo %d@%d with %d host halos across %d snapshots" %
+            (HostHaloId, isnap, len(hosts), len(snaps)))
+
+        cmh = np.array(
+            zip(np.full_like(snaps, HostHaloId), snaps, np.zeros_like(snaps)),
+            dtype=np.dtype([
+                ('HaloId', np.int32),
+                ('Snapshot', np.int32),
+                ('M200Crit', np.float32),
+            ]))
+        for i, _ in np.ndenumerate(cmh):
+            cmh[i]['M200Crit'] = np.sum(filter(lambda m: m > f * m0,\
+                hosts[hosts['Snapshot'] == cmh[i]['Snapshot']]['M200Crit']))
+
+        logging.info("Finished CMH for halo %d@%d" % (HostHaloId, isnap))
+
+        return cmh
+
+    def GetHostProfile(self, selection=None, isnap=-1):
+        """Returns normalised, binned particle positions of a FoF group.
+        """
+
+        logging.info('Retrieving profile for halos %s' % str(selection))
+        profile = self.LoadHostHalos(isnap, selection)['Profile']
+        return profile
+
+    def CalculateProfile(self, TrackId, isnap=-1, bins=None):
+        """Returns normalised particle positions of a subhalo.
+
+        If bins are provided, particle positions are binned, and a density
+        profile is returned.  Otherwise, raw positions, centered around
+        ``ComovingAveragePosition`` and normalised to ``BoundR200CritComoving``
+        are returned.
+        """
+
+        result = []
+        subhalo = self.GetSub(TrackId, isnap)
+        positions = \
+            (self.GetParticleProperties(TrackId, isnap)['ComovingPosition'] -
+             subhalo['ComovingAveragePosition'][0]) /\
+            subhalo['BoundR200CritComoving']
+
+        if bins is not None:
+            distances = np.apply_along_axis(
+                lambda x: np.sqrt(np.sum(np.power(x, 2.0))), 1, positions)
+            result = np.histogram(distances, bins=bins)
+        else:
+            result = positions
+
+        return result
+
+    def CalculateHostProfile(self, HostHaloId, isnap=-1, bins=None):
```

Comment:
better implement this outside HBTReader, since it's computing rather than reading. And different user may want different implementation (e.g., using all particles instead of FoF), so this is not of very general interest.